File: buffer/gpu/src/jvmMain/java/com/wsr/gpu/JCollection.java
package com.wsr.gpu;

class JCollection {
    public native void averageD1(long x, long result, long runtime);
    public native void averageD2(long x, int xi, int xj, int axis, long result, long runtime);
    public native void averageD3(long x, int xi, int xj, int xk, int axis, long result, long runtime);

    public native void maxD1(long x, long result, long runtime);
    public native void maxD2(long x, int xi, int xj, int axis, long result, long runtime);
    public native void maxD3(long x, int xi, int xj, int xk, int axis, long result, long runtime);

    public native void minD1(long x, long result, long runtime);
    public native void minD2(long x, int xi, int xj, int axis, long result, long runtime);
    public native void minD3(long x, int xi, int xj, int xk, int axis, long result, long runtime);

    public native void sumD1(long x, long result, long runtime);
    public native void sumD2(long x, int xi, int xj, int axis, long result, long runtime);
    public native void sumD3(long x, int xi, int xj, int xk, int axis, long result, long runtime);
}

File: buffer/cpu/src/jvmAndroidMain/kotlin/com/wsr/knist/cpu/reduction/JCollection.kt
package com.wsr.knist.cpu.reduction

import java.nio.ByteBuffer

class JCollection {
    external fun averageD1(x: ByteBuffer, result: ByteBuffer)
    external fun averageD2(x: ByteBuffer, xi: Int, xj: Int, axis: Int, result: ByteBuffer)
    external fun averageD3(x: ByteBuffer, xi: Int, xj: Int, xk: Int, axis: Int, result: ByteBuffer)
    external fun averageD4(x: ByteBuffer, xi: Int, xj: Int, xk: Int, xl: Int, axis: Int, result: ByteBuffer)

    external fun maxD1(x: ByteBuffer, result: ByteBuffer)
    external fun maxD2(x: ByteBuffer, xi: Int, xj: Int, axis: Int, result: ByteBuffer)
    external fun maxD3(x: ByteBuffer, xi: Int, xj: Int, xk: Int, axis: Int, result: ByteBuffer)
    external fun maxD4(x: ByteBuffer, xi: Int, xj: Int, xk: Int, xl: Int, axis: Int, result: ByteBuffer)

    external fun minD1(x: ByteBuffer, result: ByteBuffer)
    external fun minD2(x: ByteBuffer, xi: Int, xj: Int, axis: Int, result: ByteBuffer)
    external fun minD3(x: ByteBuffer, xi: Int, xj: Int, xk: Int, axis: Int, result: ByteBuffer)
    external fun minD4(x: ByteBuffer, xi: Int, xj: Int, xk: Int, xl: Int, axis: Int, result: ByteBuffer)

    external fun sumD1(x: ByteBuffer, result: ByteBuffer)
    external fun sumD2(x: ByteBuffer, xi: Int, xj: Int, axis: Int, result: ByteBuffer)
    external fun sumD3(x: ByteBuffer, xi: Int, xj: Int, xk: Int, axis: Int, result: ByteBuffer)
    external fun sumD4(x: ByteBuffer, xi: Int, xj: Int, xk: Int, xl: Int, axis: Int, result: ByteBuffer)

    external fun maxIndexD1(x: ByteBuffer, result: ByteBuffer)
    external fun maxIndexD2(x: ByteBuffer, xi: Int, xj: Int, axis: Int, result: ByteBuffer)
    external fun maxIndexD3(x: ByteBuffer, xi: Int, xj: Int, xk: Int, axis: Int, result: ByteBuffer)
}
